### PR TITLE
S3 compare-and-exchange implementation

### PR DESCRIPTION
**Pull request by DaveCTurner**
Adds an implementation of `compareAndExchangeRegister` to `S3BlobContainer`.

### CODE REVIEW

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java [hunk 1]
```diff
@@ -610,18 +623,211 @@ static Tuple<Long, Long> numberOfMultiparts(final long totalSize, final long par
...
+                )
+            ) {
+                if (currentUploads.size() > 1) {
+                    // This is a small optimization to improve the liveness properties of this algorithm.
+                    //
+                    // When there are multiple competing updates, we order them by upload id and the first one tries to cancel the competing
+                    // updates in order to make progress. To avoid liveness issues when the winner fails, the rest wait based on their
+                    // upload_id-based position and try to make progress.
+
+                    var delayListener = listeners.acquire();
```
Comment (fcofdez) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:751]: I think there's a risk of releasing all the `listeners` refs before the multipart upload tasks are dispatched into the `SNAPSHOT` thread pool if?

**Review by fcofdez — COMMENTED** [review]
It looks mostly good to me, I left a comment about the ref counting in the abort case.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java (hunk 1 above)
Comment (DaveCTurner) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:751]: We `.acquire()` a listener for each upload while constructing the `ActionRunnable` to be dispatched, and that all happens on this thread before completing `delayListener`.

Comment (fcofdez) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:751]: You're right, thanks for the clarification 👍

**Review by fcofdez — APPROVED** [review]
LGTM

**Comment by DaveCTurner** [comment]
Thanks @fcofdez!

@elasticmachine update branch

**Pull request merged**